runtime/ops/fs_events.rs

```diff
@@ -93,43 +92,90 @@ impl From<NotifyEvent> for FsEvent {
   }
 }
 
-#[derive(Deserialize)]
-pub struct OpenArgs {
-  recursive: bool,
+type WatchSender = (Vec<String>, mpsc::Sender<Result<FsEvent, AnyError>>);
+
+struct WatcherState {
+  senders: Arc<Mutex<Vec<WatchSender>>>,
+  watcher: RecommendedWatcher,
+}
+
+fn starts_with_canonicalized(path: &Path, prefix: &str) -> bool {
+  #[allow(clippy::disallowed_methods)]
+  let path = path.canonicalize().ok();
+  #[allow(clippy::disallowed_methods)]
+  let prefix = std::fs::canonicalize(prefix).ok();
+  match (path, prefix) {
+    (Some(path), Some(prefix)) => path.starts_with(prefix),
+    _ => false,
+  }
+}
+
+fn start_watcher(
+  state: &mut OpState,
   paths: Vec<String>,
+  sender: mpsc::Sender<Result<FsEvent, AnyError>>,
+) -> Result<(), AnyError> {
+  if let Some(watcher) = state.try_borrow_mut::<WatcherState>() {
+    watcher.senders.lock().push((paths, sender));
+    return Ok(());
+  }
+
+  let senders = Arc::new(Mutex::new(vec![(paths, sender)]));
+
+  let sender_clone = senders.clone();
+  let watcher: RecommendedWatcher = Watcher::new(
+    move |res: Result<NotifyEvent, NotifyError>| {
+      let res2 = res.map(FsEvent::from).map_err(AnyError::from);
+      for (paths, sender) in sender_clone.lock().iter() {
+        // Ignore result, if send failed it means that watcher was already closed,
+        // but not all messages have been flushed.
+
+        // Only send the event if the path matches one of the paths that the user is watching
+        if let Ok(event) = &res2 {
+          if paths.iter().any(|path| {
+            event.paths.iter().any(|event_path| {
+              same_file::is_same_file(event_path, path).unwrap_or(false)
+                || starts_with_canonicalized(event_path, path)
+            })
```

Comment:
Can we be smarter here? Iterating over all paths specified for each watcher can be costly. Maybe a hash set would work better here to try and match file path directly?

Reply:
Two watchers can be watching the same file/directory so HashSet won't work.